### ISSUE TITLE
Levenshtein

### ISSUE THREAD

**Issue by softwx** [wolfgarbe/SymSpell#24]
Wolf,
I've updated the GitHub project that has my Levenshtein function. I've updated it to support using it via an instantiated class. Using it this way, it uses almost no memory, even temporarily, so it's memory impact is extremely negligible. I also made it as a nuget package, that is packaged as a .net standard 1.0 library. If you're interested in pulling it into symspell, let me know, and I can do that and submit a pull request. It would be cleaner I think, if the source was pulled into the symspell project, rather than create the package dependency. Before doing it though, it might be good if I had a better idea about how sysspell is used by you and other folks as far as multi-threading is concerned. The levenshtein distance can be done via an instantiated class, but the distance function itself is not threadsafe. There is also a static version of the function, that is only a tiny bit slower, but it loses the memory consumption advantage of the instantiated version. 
Steve

**Comment by wolfgarbe** [comment]
Steve,
For threading SymSpell I see three basic use cases:

1. A single SymSpell instance runs in a single thread. Lookups can be consumed by a single thread or in parallel by multiple threads (e.g. parallel queries from different users to a web search engine or a data base).  I think Lookup should be thread safe, but LoadDictionary, CreateDictionary, and CreateDictionaryEntry have to be synchronized by locks.

2. Multiple SymSpell instances run on multiple threads, each instance on their own thread. 
This leads to high memory consumption with each SymSpell instance having its own dictionary in RAM. 
Therefore it makes only sense if different dictionaries for different languages, domains or customer-specific data/vocabulary are required in parallel. 

3. Utilize multiple threads to speed up a single Lookup. If SymSpell is the bottleneck (and there are no parallel SymSpell lookups which would anyway utilize all processor cores like in case 1)
then we could reduce the response time by splitting a single Lookups into parallel sub-tasks running on multiple threads distributed over multiple processor cores. 
E.g. the recursions for the delete generations and lookups could be run in parallel. This is currently not implemented.

I would like to support all three cases, without forcing the user to adapt their architecture to the spelling correction component. 

What do you think, how would the static vs the instantiated class of your Levenshtein function would fit best into these cases? It would be great if you could submit a pull request to pull your source into the SymSpell project, rather than create the package dependency.

Wolf

**Comment by softwx** [comment]
See new pull request #27

**Comment by wolfgarbe** [comment]
I was thinking about how we could utilize the similarity between the multiple suggestions of a single input term to speed up the levenshtein calculation (something similar to http://stevehanov.ca/blog/index.php?id=114, but he expects the suggestions in alphabetical order). What do you think?

**Comment by softwx** [comment]
A trie could be a compact way to store the entire dictionary. I would be curious to compare it, but something to remember is that the Damerau and Levenshtein algorithms in SoftWx.Match are pretty heavily optimized. When comparing two words with a maxDistance given (which is how symspell uses it) the time complexity is O(2 * maxDistance * m), which is essentially O(m), i.e. linear time, not O(m*n) like the standard Levenshtein (where m is the length of the shorter string, and n is the length of the longer string). Also, shared suffixes and prefixes of the two words are quickly identified and ignored from further consideration, and the algorithm exits early as soon as it's known that the maxDistance can't be met. So, for example comparing "portent" and "potient", what actually makes it to the core levenshtein loops is "rt" and "ti".
So, I think the trie is not so much replacing the way candidates are evaluated against the input, as it is replacing the way the dictionary is stored and candidates determined and evaluated. That is, unless I'm misunderstanding how you were thinking of applying it.

**Comment by wolfgarbe** [comment]
He uses a trie, but that's not the core of the idea to speed the calculation up. 
He uses the prefix trie only to store similar words together and process them in alphabetical order.
We know already in advance that all suggestion candidates of a single delete are similar - therefore we don't necessarely need the trie.
But his core idea is that for similar words it's not necessary to create the entire matrix from scratch for each suggestion. Only those lines where the suggestions differ have to be recreated!

Skipping a shared prefix and suffix helps, but does not utilize the full potential of similarity:
input: btest
suggestion1: testa
suggestion2: testb
suggestion3: testbc

They neither share a prefix nor a suffix, but they are still similar!

**Comment by softwx** [comment]
After looking at that and having a quick think, I believe a proof of concept could be done pretty simply by adding memoization to the Damerau/Levenshtein algorithms, implemented simply using a Dictionary. While not the most memory efficient, it should allow fast implementation, and give an idea if it's a productive direction. I don't see a need for the candidates to be evaluated in sorted order. I may be able to throw it together in a few days.

**Comment by wolfgarbe** [comment]
Sounds good!

**Comment by softwx** [comment]
I did some tests, and here's what I found. Because the SoftWx Damerau and Levenshtein functions are heavily optimized, implementing something like this memoize idea comes at the expense of some of the optimizations as they are incompatible. The author of that article saw dramatic improvements because he is comparing to the unoptimized standard version of Levenshtein. It would almost be better to implement it from scratch against the standard Damerau algorithm, and then try to optimize the result. But I'm skeptical that would produce much improvement, if any. Here's why. Your core symspell already narrows the set of candidates down to those that are going to be pretty similar. The candidates that are left are pretty likely to have at least some shared prefix or suffix in many cases, something that the SoftWx functions are optimized for. Also, you use relatively narrow max distances, which make a huge difference, something he wasn't considering. Even if there were no shared prefix or suffix, take the following example: "lhanton" compared to "phantom". With a maxDistance of 3, the SoftWx Damerau takes 178 nanoseconds. To implement that article's idea, means putting a fair amount of new overhead into the distance function. At the start it will have to do lookups to see if some part of the prefix has already been memorized, and then, as the characters of the main string are iterated over in the internal distance loop, its state will need to be memorized for the benefit of subsequent calls of the distance function against other candidates.  Two options for the memoize structure would be Dictionary for hashed lookups, or simple arrays which would have to be searched linearly, which might not be bad since they will be small arrays. For Dictionary, it's 16 nanoseconds to do a lookup. That adds some overhead, but since you look for the longest memorized prefix, starting with just the first character, and stop when you reach the length you fail to find one, you probably only do that one to maybe a few times. What's worse is adding to the memorized structure. Dictionary Add takes 26 nanoseconds. You will do that from the length of any shared prefix you found, out to the end of the string. So that will generally be the length of the string, less a few characters if you happened to start with some already memorized prefix. You can see, that overhead can add up pretty fast compared to 178 nanoseconds it takes now. In the case of "phantom", if there was already a memorized prefix of "pha", that's 4 * 16ns = 64ns to look up the shared prefix (the 4th one being the failed lookup that stops looking for more). Then you are left to do the regular damerau of "ntom" (so you saved about a little less than half the damerau looping you'd otherwise have to have done). Also, you have to memoize "phan", "phant", "phantom", and "phantom. That's 4 * 26ns =  106ns of additional overhead. The opportunity for it to beat the current algorithm is even worse if there is a shared prefix. For example, having just two share characters in the suffix, as in "lhandom" to "phantom" reduces the current Damerau algorithm time from 178ns down to 111ns.
There are several things I see making the article's idea less fruitful for symspell: use of max distance reduces much of the opportunity for it to shave off time, a chunk of the time it would save will be consumed by the added overhead of maintaining the structure it uses, some of the beneficial optimizations in the current Damerau would not be able to be used, it would be generate more temporary object creation and memory use, it would be more complicated to implement in Damerau than in Levenshtein.
It is an interesting idea. But I think to really do it would mean starting with the standard unoptimized Damerau, get that working, and then trying to do some further optimizations on that. That's probably going to take more time that I have, particularly since I'm not convinced the end result will be an improvement. It may be possible to take some of the ideas though, and use them to make some optimizations to the existing algorithm that aren't incompatible with the existing optimizations.

**Comment by wolfgarbe** [comment]
Steve, thanks a lot for your detailed analysis. I think you are right with your reasoning. 
Looks like your algorithm is just too good to be be further optimized ;-)

Wolf

**Comment by softwx** [comment]
I still have a few ideas left to try.
